aten/src/ATen/native/cuda/FractionalMaxPool2d.cu

```diff
@@ -0,0 +1,375 @@
+      if (PoolSizeWStatic == -1) {
+        for (int w = poolW; w < poolW + poolSizeW; ++w) {
+          scalar_t val = input[batch][plane][h][w];
+          // for consistency with THNN, favor the first max
+          if (val > maxVal) {
+            maxIndex = h * input.size(3) + w;
+            maxVal = val;
+          }
+        }
+      } else {
+#pragma unroll
+        for (int i = 0; i < PoolSizeWStatic; ++i) {
+          int w = i + poolW;
+          scalar_t val = input[batch][plane][h][w];
+          // for consistency with THNN, favor the first max
+          if (val > maxVal) {
+            maxIndex = h * input.size(3) + w;
+            maxVal = val;
+          }
+        }
+      }
+    }
+
+    assert(maxVal != at::numeric_limits<scalar_t>::lowest());
+    assert(maxIndex != -1);
+
+    indices[batch][plane][outputH][outputW] = maxIndex;
+    output[batch][plane][outputH][outputW] = maxVal;
+  }
+}
+
+template <typename scalar_t>
+__global__ void fractional_max_pool2d_backward_out_cuda_frame(
+  PackedTensorAccessor<scalar_t, 4> gradInput,
+  PackedTensorAccessor<scalar_t, 4> gradOutput,
+  PackedTensorAccessor<int64_t, 4> indices) {
+  // Output (h, w) point that this thread is responsible for
+  int ourOutputPoint = threadIdx.x + blockIdx.x * blockDim.x;
+  int plane = blockIdx.y;
+  int batch = blockIdx.z;
+
+  // Each thread generates a specific output point
+  if (ourOutputPoint < gradOutput.size(2) *
+    gradOutput.size(3)) {
+    int outputW = ourOutputPoint % gradOutput.size(3);
+    int outputH = ourOutputPoint / gradOutput.size(3);
+
+    int index = indices[batch][plane][outputH][outputW];
+    assert(index >= 0);
+    int inputW = index % gradInput.size(3);
+    int inputH = index / gradInput.size(3);
+    assert(inputH < gradInput.size(2));
+
+    atomicAdd(
+      &gradInput[batch][plane][inputH][inputW],
+      gradOutput[batch][plane][outputH][outputW]
+    );
+  }
+}
+
+void fractional_max_pool2d_out_cuda_template(
+  Tensor & output,
+  Tensor& indices,
+  const Tensor& input,
+  IntList pool_size,
+  IntList output_size,
+  const Tensor& randomSamples) {
+  int planeDim = 0;
+  int dimh = 1;
+  int dimw = 2;
+  int numBatch = 1;
+
+  int ndims = input.ndimension();
+  for (int i = 0; i < ndims; i++) {
+     AT_CHECK(input.size(i) > 0,
+       "fractional_max_pool2d(): expected input to have non-empty spatial dimensions, "
+       "but input has sizes ", input.sizes(), " with dimension ", i, " being "
+       "empty");
+   }
+
+   AT_CHECK((ndims == 3 || ndims == 4),
+     "non-empty 3D or 4D (batch mode) tensor expected for input");
+
+  if (ndims == 4) {
+    numBatch = input.size(0);
+    planeDim++;
+    dimh++;
+    dimw++;
+  }
+
+  /* sizes */
+  int numPlanes = input.size(planeDim);
+  int inputH = input.size(dimh);
+  int inputW = input.size(dimw);
+
+  int outputH = output_size[0];
+  int outputW = output_size[1];
+  int poolSizeH = pool_size[0];
+  int poolSizeW = pool_size[1];
```

Comment:
This doesn't match the original code...

```
void THNN_(SpatialFractionalMaxPooling_updateOutput)(
           THCState *state,
           THCTensor *input,
           THCTensor *output,
           int outputW, int outputH,
           int poolSizeW, int poolSizeH,
           THCIndexTensor *indices,
           THCTensor *randomSamples)
```

...but I think I agree that the new code is correct and the old code is wrong

Reply:
The old code is correct too, weirdly enough; see https://github.com/pytorch/pytorch/pull/15531#discussion_r245895241

Reply:
For this case I followed the doc in torch/nn/modules/pooling.py